부산대 BE_정윤성 2단계 - 관리자 화면

dev-yunseong opened the pull request:
> 관리자 페이지 구현, 1차 리뷰 반영을 했습니다.

* 관리자 페이지를 thymleaf를 통해 서버 사이드 렌더링으로 구현을 했습니다.
  * Modal을 사용해서 수정, 생성에서 재사용하도록 구현했습니다.
    * title은 동적으로 적용해서 구분 가능하도록 했습니다.
  * css 파일을 table, modal, base로 분리하여 가독성과 유지보수성을 높였습니다.
* 1차 리뷰 반영
  *  멱등성을 고려해 delete 에서 없는 id로 삭제 요청을 한다고 해도 no-content로 반응하지 않도록 수정했습니다.
  * 생략 가능하거나 필요 없는 Annotation을 삭제 했습니다.

dev-yunseong committed:
doc README.md - 구현할 기능 목록 최신화

dev-yunseong committed:
feature 관리자 페이지 구현 - table을 통해 product를 출력하도록 [thymleaf를 통해 서버 사이드 렌더링], Product 정보를 받기 위한 modal 창 구현

dev-yunseong committed:
feature 제품 관리 기능 구현 - modal을 통해서 제품 정보를 받아서, post, put, delete 기능 구현

dev-yunseong committed:
feature style 정의 - 관리자 화면을 더욱 보기 좋게 css로 style 정의

dev-yunseong committed:
chore application.yml - yml 형식으로 수정, thymeleaf 설정 추가

dev-yunseong committed:
feature AdminController - 관리자 페이지 controller 구현

dev-yunseong committed:
chore 1차 리뷰 반영 - interface에 불필요한 Repository annotation 제거, 생략 가능한 annotation 제거, 멱등성을 고려한 로직 수정

jaenyeong commented on src/main/java/gift/controller/AdminController.java:
```diff
@@ -0,0 +1,22 @@
+        this.productService = productService;
+    }
+
+    @GetMapping("/admin/products-management")
+    public String adminDashboard(Model model) {
+        model.addAttribute("products", productService.findAllProducts());
+        return "products-management";
+    }
```
❓ 
해당 URL도 충분히 직관적이지만, 조금 더 간결한 URL을 설계해 보시면 어떨까요?
제 개인적인 생각은 admin과 management 용어가 꼭 두개가 함께 사용될 필요가 있을까라는 생각도 드네요 ㅎㅎ

앞으로 많은 API URL을 설계하시게 될텐데 직관적이고 간결하게 그리고 확장성 있는 URL 들을 설계해 보시는 연습을 해보시면 좋겠습니다 ㅎㅎ

dev-yunseong replied:
github url을 참고해서 
"admin/products" 로 수정했습니다.
이후 "admin/users"와 같은 식으로 확장 가능할 것 같고 더욱 읽기 쉬워진 것 같습니다.

jaenyeong commented on src/main/java/gift/repository/MemoryProductRepository.java:
```diff
@@ -30,9 +29,6 @@ public Product findProductById(Long id) {
 
     @Override
     public void deleteProductById(Long id) {
-        if (!products.containsKey(id)) {
-            throw new NotFoundByIdException("Product with id " + id + " does not exist.");
-        }
```
👍 
찾아보는 행위를 줄이면서 코드가 간결해졌네요 ㅎㅎ
프로그래밍 언어의 최신 문법들을 사용하면서 가독성 좋은 코드를 작성하는 것도 중요하지만, 애초에 번거로운 로직을 덜어낼 수 있는지도 함께 고민하는 습관을 가지시면 좋을 것 같네요~

jaenyeong commented on src/main/resources/templates/products-management.html:
```diff
@@ -0,0 +1,53 @@
+                <label>상품명: <input type="text" name="name" required></label><br><br>
+                <label>가격: <input type="number" name="price" required></label><br><br>
+                <label>사진: <input type="text" name="imageUrl" required></label><br><br>
+                <button type="submit">제출</button>
```
🔴 
Add Product 버튼으로 데이터를 생성할 때, 
가격 부분에 `1111111111111111111111111111111` 또는 `12121212122121212121212121212121212` 처럼 입력하게 되면 아래 에러가 발생됩니다!

```text
Resolved [org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `java.lang.Integer` from String "12121212122121212121212121212121212": not a valid `java.lang.Integer` value]
```

어떻게 해결하면 좋을까요?
그리고 간단한 과제라 생각지 못하셨을 수 있는데요~
지금 상품 dto의 price 필드는 Integer인데, 위처럼 파라미터가 문자열로 인식하게 되지 않았더라면 이 값의 범위를 넘는 값이 들어왔을 때 어떻게 될까요? ㅎㅎ 또 이런 처리는 어떻게 하면 좋을까요?

dev-yunseong replied:
먼저 유저와의 반응성 측면에서 먼저 프론트에서 max, min, step을 설정했습니다.
또한 클라이언트에서 뭔가 비정형적인 요청을 보낼 때를 대비해서, 라우터에서 처리를 해보려고 했는데 Json 파싱 구간에서 예외가 나기 전에 그 validation을 할 순 없을 것 같아서, HttpMessageNotReadableException에 대한 Exception Handler를 추가 했습니다.

jaenyeong replied:
❓ 
> 라우터에서 처리를 해보려고 했는데 Json 파싱 구간에서

혹시 여기서 말씀하신 라우터가 정확히 어디를 얘기하시는 걸까요?

dev-yunseong replied:
Servlet에서 Controller로 오는 구간 말하는 것이었습니다!

jaenyeong reviewed (changes requested):
안녕하세요 윤성님~
조금 코멘트를 남겼는데요! 사실 프론트 쪽은 크게 리뷰 드릴 영역이 아니라고 생각하였는데 저도 모르게 버그를 발견해서 버그 수정 요청을 드립니다!

읽기 쉬운 코드, 유연한 코드, 성능 좋은 코드를 작성하는 것도 중요하지만 요구사항에 맞는 코드, 버그가 없는 코드도 못지 않게 중요하다는 것을 늘 생각해 주셨으면 좋겠습니다 ㅎㅎㅎ

dev-yunseong committed:
chore 2차 리뷰 반영 -
1. github url을 참고해서 url를 만들어봤습니다. admin/users 와 같은 식으로 확장 가능하고, 읽기 편하게 수정했습니다.
2. Int 범위를 넘어서를 값을 넘었을 때 예외처리를 위해 먼저 프론트에서 제한을 걸었습니다. 프론트에서 제한 만으로 충분하지 않다고 생각이 되어서 Controller에 ExceptionHandler까지 추가했습니다.

dev-yunseong commented:
안녕하십니까? 좋은 리뷰 감사드립니다!

dev-yunseong commented:
리뷰 해주신 것을 한번 반영 해봤습니다. 한번 더 리뷰 해주시면 감사하겠습니다~

dev-yunseong commented:
> 리뷰 해주신 것을 한번 반영 해봤습니다. 한번 더 리뷰 해주시면 감사하겠습니다~

Next step에서도 다시 리뷰 요청 해놓겠습니다!

jaenyeong commented on src/main/java/gift/controller/AdminController.java:
```diff
@@ -0,0 +1,22 @@
+    @GetMapping("/admin/products")
+    public String adminDashboard(Model model) {
+        model.addAttribute("products", productService.findAllProducts());
+        return "products-management";
```
🟢 
API와 반환될 페이지명(또는 파일명)이 API URL(path)와 꼭 똑같을 필요는 없지만 운영상 이점을 위해 어느 정도는 비슷하게 네이밍하는 것도 고려해보시길 권해드려요 ㅎㅎ 물론 크리티컬하진 않은 부분이지만 이런 부분은 항상 협업하는 분들과 커뮤니케이션을 통해 풀어나가야 하는 부분일 수 있어요~

jaenyeong commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -57,4 +56,9 @@ public ResponseEntity<String> handleNotFoundByIdException(NotFoundByIdException
                 .body("Not Found by ID: " + e.getMessage());
     }
+
+    @ExceptionHandler(HttpMessageNotReadableException.class)
+    public ResponseEntity<String> handleJsonParseError(HttpMessageNotReadableException e) {
+        return ResponseEntity.badRequest().body("Invalid Request: " + e.getMessage());
+    }
```
👍 
특정 예외에 대한 핸들러를 구현하여 처리한 것은 좋은 생각 같아요~

다만 현업 입장에서 조금 팁을 드리자면, 특정 에러가 발생했을때 클라이언트에게 세부적인 메시지를 고스란히 전달하지는 않는 편입니다~
이유는 무엇일까요? 이런 경우 보통 현업에선 어떻게 처리할까?도 고민해 보시면 좋을 것 같네요 ㅎㅎ

아래와 같은 것들을 살펴 보시면 좋겠습니다!
1. 에러 메시지를 정확히 알아야 하는 이유는 무엇일까?
2. 이 메시지를 필요로 하는 건 어디일까? 또 어디에 기록이 남아야 할까?
3. 클라이언트에서 예외 핸들링을 위해 세부 메시지를 꼭 알아야만 할까?

jaenyeong commented on src/main/resources/static/js/products-management.js:
```diff
@@ -0,0 +1,61 @@
+    sendProductData('http://localhost:8080/api/products', 'POST');
+}
+
+function updateProduct(id) {
+    sendProductData(`http://localhost:8080/api/products/${id}`, 'PUT');
+}
+
+function deleteProduct(productId) {
+    if (productId) {
+        fetch(`http://localhost:8080/api/products/${productId}`, {
+            method: 'DELETE'
+        }).then(() => {
+            location.reload();
+        });
+    }
+}
```
🟢 
이 부분은 프론트/백엔드 상관없는 이야기인데요! updateProduct 함수의 파라미터명과 deleteProduct 함수의 파라미터명이 다른데, 실질적으로 같은 productId를 의미하죠~ 이런 경우 가급적 네이밍을 통일해 주시면 인적장애(사람의 실수)를 방지하는데 도움이 되기도 합니다 ㅎㅎ

jaenyeong reviewed (approved):
고생하셨습니다~

앞선 리뷰로 인해 price 사이즈에 대한 것들이 수정 되었는데, `README.md` 파일의 요구사항은 수정된 것이 없어 보여요~ 이런 것들도 꼼꼼히 기록을 남기셔야 의도된 요구사항인지를 파악하는데 도움이 될 것 같아서 항상 이런 요구사항을 명확히 하시는 습관을 가지시면 좋겠습니다!

그리고 제 개인적으로는 화면사이드는 해당 백엔드 교육 과정의 학습 부분에서 크게 중요하게 생각하지 않아서 별다른 피드백을 드리진 않았습니다~ (하지만 질문이 있다면 편하게 남겨 주세요 ㅎㅎ)

다만 놓칠 수 있는 부분들을 코멘트 남겨 놓았고, 역시나 꼭 답변을 하거나 제 조언을 받아들이진 않아도 됩니다! 다만 본인만의 생각을 조금 더 되돌아 보는 포인트가 되셨으면 하고, 본인의 생각을 남에게 설득 시킬 수 있도록 논리적으로 정리하면 좋을 것 같아요 ㅎㅎ

dev-yunseong commented:
> 고생하셨습니다~
> 
> 
> 
> 앞선 리뷰로 인해 price 사이즈에 대한 것들이 수정 되었는데, `README.md` 파일의 요구사항은 수정된 것이 없어 보여요~ 이런 것들도 꼼꼼히 기록을 남기셔야 의도된 요구사항인지를 파악하는데 도움이 될 것 같아서 항상 이런 요구사항을 명확히 하시는 습관을 가지시면 좋겠습니다!
> 
> 
> 
> 그리고 제 개인적으로는 화면사이드는 해당 백엔드 교육 과정의 학습 부분에서 크게 중요하게 생각하지 않아서 별다른 피드백을 드리진 않았습니다~ (하지만 질문이 있다면 편하게 남겨 주세요 ㅎㅎ)
> 
> 
> 
> 다만 놓칠 수 있는 부분들을 코멘트 남겨 놓았고, 역시나 꼭 답변을 하거나 제 조언을 받아들이진 않아도 됩니다! 다만 본인만의 생각을 조금 더 되돌아 보는 포인트가 되셨으면 하고, 본인의 생각을 남에게 설득 시킬 수 있도록 논리적으로 정리하면 좋을 것 같아요 ㅎㅎ

리뷰 감사드립니다!
